proposals/20210525-feast_example_gen.md

```diff
@@ -0,0 +1,75 @@
+
+**Your company/organization:** Apple
+
+**Project name:** ExampleGen for Feast Feature Store
+
+## Project Description
+
+This proposal is for a new component to enable connecting to and reading from feature stores via a simple and unified API. The first iteration will only support Feast [1], but we will add support for other popular feature stores over time. Similarly, we will restrict v1 of our library to only read from feature stores- we will evaluate supporting writes later.
+
+## Project Category
+Component
+
+## Project Use-Case(s)
+
+Feature stores such as Feast are becoming increasingly popular as (managed) platforms to manage all feature-related information. A separate feature store has several benefits including (but not limited to) the following [2]:
+
+1. Feature sharing and reuse
+2. Serving features at scale
+3. Consistency between training and serving (inference)
+4. Point-in-time correctness
+5. Data quality and validation
+
+Different organizations may use different feature stores in their ML stack, each with its own APIs for reading online and historical features. This is a significant barrier to users as they will have to learn to use a different set of APIs for each feature store. Furthermore, different feature stores may return different feature vector representations. The proposed component will solve these issues:
+
+1. It will provide a unified API across different feature stores. It will hide the mechanics of reading from different feature stores from the user.
+2. It will always return the feature vector as a tensor. It will convert different representations of the feature vector to a tensor transparently from the user.
+
+**Disclaimers**
+
+Our organizations currently do not use the project and they have no plans to use it once it is implemented as far as we know.
+
+There is no potential impact, overlap, or synergies with other projects. Dependencies are outlined below.
+
+## Project Implementation
+
+* At a high level, there are 3 major APIs [3]:
+1. Connect to a remote Feast instance. The connection string could be part of a general procedure to configure the feature store using a JSON/YAML file.
+2. Provide a wrapper for `get_online_features`. Feast returns an `OnlineResponse` [5], which will be converted to a tensor.
+3. Provide a wrapper for `get_historical_features`. Feast returns a `RetrievalJob` [6], which will be used to materialize the results and convert them to a tensor.
+
+* Important Considerations
+1. We want to generalize the implementation as much as possible to make it straightforward to add support for other feature stores besides Feast in the future. To this end, it would be useful to generalize the notion of Feast's entity types [4].
+
+2. Today Feast only supports BigQuery queries for `get_historical_features`. We should parameterize the query so that users can use other offline data stores in the future. We should propagate the error message for bad queries.
```

Comment:
Please see this [RFC](https://docs.google.com/document/d/1MD0aS2_hGzd1tJ7DNjE3NgEtcuekh3O06OeQ9aavylY/edit#) to add support for different backends (compute, storage) to Feast. We are also working on adding more offline stores right now, specifically Redshift (with Snowflake planned).

Reply:
This is great! I see tremendous value in users interfacing through a single feature store (Feast) for all things feature-related instead of having to use separate `ExampleGen`s for different datastores.